openshift volume attach/detach

chitra607 opened the pull request.

chitra607 committed:
openshift volume attach/detach

agrare commented:
@chitra607 does this PR replace https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/302 ?

agrare commented on app/models/manageiq/providers/kubevirt/infra_manager/vm/operations.rb:
```diff
@@ -5,6 +5,7 @@ module ManageIQ::Providers::Kubevirt::InfraManager::Vm::Operations
   include Power
   include Snapshot
   include Guest
+  include InfraVolume
```
Minor, but in terms of naming I think `Disk` would be a better module name as that is in line with our nomenclature

agrare commented on app/models/manageiq/providers/kubevirt/infra_manager/vm/operations/infra_volume.rb:
```diff
@@ -0,0 +1,157 @@
+            :storage => volume_size
+          }
+        },
+        :storageClassName => "lvms-vg1"
```
This looks like something we'd want to let the user select

agrare commented:
@chitra607 the existing methods that we expose on the VmOrTemplate class for InfraManager type vms consists of:
* `add_disk(disk_name, disk_size_mb, options = {})`
* `remove_disk(disk_name, options = {})`
* `resize_disk(disk_name, disk_size_mb, options = {})`
* `reconfigure(spec)`

Anything we add should conform to these methods.

agrare commented on app/models/manageiq/providers/kubevirt/infra_manager/vm/operations/infra_volume.rb:
```diff
@@ -0,0 +1,157 @@
+module ManageIQ::Providers::Kubevirt::InfraManager::Vm::Operations::InfraVolume
+  def raw_attach_volume(vm, pvc_name, volume_name = nil)
```
You have `vm` as a param here but this is an instance method in the vm class so you can just use `self` for anything that you need, e.g. instead of `ems = vm.ext_management_system` you can just use `ext_management_system` and that is effectively `self.ext_management_system`

Dhamo1107 commented on app/models/manageiq/providers/kubevirt/infra_manager/vm/operations/infra_volume.rb:
```diff
@@ -0,0 +1,157 @@
+
+  def create_pvc(vm, volume_name, volume_size)
+    ems = vm.ext_management_system
+    kubevirt = ems.parent_manager.connect(
+      :service => "kubernetes",
+      :path    => "/api",
+      :version => "v1"
+    )
```
Hi @agrare,
Thanks for the feedback!

I tried using `ems.kubeclient` instead of `parent_manager.connect(...)`, but ran into an issue.


```ruby
ems = ManageIQ::Providers::Kubevirt::InfraManager.first  
ems.kubeclient  
# => NoMethodError: undefined method `kubeclient' for ManageIQ::Providers::Kubevirt::InfraManager
```

* Is the `kubeclient` method expected to exist on the KubeVirt provider?
* Or is this something that needs to be implemented?
* Or is there a recommended way to access kubeclients in the KubeVirt provider?

Thanks!

Dhamo1107 replied:
@agrare, we tried that in the Radjabov version, but the `kubeclient `method is not present in `infra_manager.rb`. It exists only in the master branch, which is the source of confusion

agrare replied:
That's probably why kubevirt-sdk-ruby wasn't loading as we just recently started using that

chitra607 committed:
Rename infra_volume to disk and update UI form

chitra607 commented:
Thanks @agrare  for the review.  
I have updated the implementation in this repo:

- Updated storageClassName so user can select it.  
- Removed the extra vm param and switched to using self method.
- Renamed the module from InfraVolume to Disk to match the existing naming convention.

Pushed the changes here:  https://github.com/ManageIQ/manageiq/commit/66aa5db919f2195f6edc0f9e90ec817e97662516
